### PR TITLE
Add error support when not enough utxo is met upon multioutput …

### PR DESCRIPTION
**Pull request by paweljakubas**
#190 

<p align="right">#</p>

# Overview

- [x] Integration test illustrating the following scenarios were added:
- [x] A multi-output transaction fails because there's not enough UTxO to cover for all outputs
- [x] A multi-output transaction succeeds
- [x] The API returns a more meaningful error when the case for (failing) multi-output is encountered, suggesting the user to have a look at /api/v1/wallets/{walletId}/statistics/utxo

# Comments

### CODE REVIEW

File: src/Cardano/Wallet/API/V1/Types.hs [hunk 1]
```diff
@@ -2409,6 +2410,28 @@ instance FromJSON ErrNotEnoughMoney where
...
+    build = \case
+        ErrUtxoNotEnoughFragmentedBy _ ->
+             bprint "Not enough utxos to accomodate all outputs of the transaction"
+
+instance ToJSON ErrUtxoNotEnoughFragmented where
+    toJSON = \case
+        (ErrUtxoNotEnoughFragmentedBy missingUtxos) -> object
+            [ "msg"          .= String "Utxo is not enough fragmented to handle the number of outputs of this transaction. see /api/v1/wallets/{walletId}/statistics/utxos for more information"
+            , "missingUtxos" .= missingUtxos
+            ]
```
Comment (KtorZ) [inline src/Cardano/Wallet/API/V1/Types.hs]: API errors are kinda... complicated at the moment but we do want to derive most things generically. I'd suggest to have:

```hs
data ErrUtxoNotEnoughFragmented = ErrUtxoNotEnoughFragmented
    { missingUtxos :: Int
    , msg :: Text
    }
```

And, provide a corresponding message as text in the `ReifyError`. This way, we can simply derive the ToJSON and FromJSON instance generically and don't need this manual wiring.

Then, make `details` the `diagnosticKey` such that it will generate a corresponding JSEnd error.

File: src/Cardano/Wallet/Kernel/Transactions.hs [hunk 2]
```diff
@@ -216,6 +233,10 @@ newUnsignedTransaction ActiveWallet{..} options accountId payees = runExceptT $
     availableUtxo <- withExceptT NewTransactionUnknownAccount $ exceptT $
                        currentAvailableUtxo snapshot accountId
 
+
+    _ <- withExceptT NewTransactionNotEnoughUtxoFragmentation $ exceptT $
+           determineCorrectUtxoFragmentation payees availableUtxo
```
Comment (KtorZ) [inline src/Cardano/Wallet/Kernel/Transactions.hs]: I'll suggest to change the signature of `determineCorrectUxoFragmentation` to `checkUtxoFragmentation :: NonEmpty (Address, Coin) -> Utxo -> Either NumberOfMissingUtxos ()` 

There's no need to return something here and the check is really just about stopping the computation or, moving forward.

File: test/integration/Test/Integration/Scenario/Transactions.hs [hunk 3]
```diff
@@ -4,12 +4,116 @@ module Test.Integration.Scenario.Transactions
...
+            & initialCoins .~ [500000, 500000]
+
+
+        randomBackPhrase1 <- liftIO $ generate arbitrary :: Scenario Context  IO Client.BackupPhrase
+        wallet1 <- successfulRequest $ Client.postWallet $- NewWallet
+                      randomBackPhrase1
+                      noSpendingPassword
+                      NormalAssurance
+                      "NewWallet1"
+                      CreateWallet
```
Comment (KtorZ) [inline test/integration/Test/Integration/Scenario/Transactions.hs]: Please use the fixture for that. Fixture is used to create an initial wallet without the hassle of generating all the bits yourself. Wallets come default accounts `defaultAccountId` that you can use in the scenario. So basically, the lines above can just be replaced with:

```hs
fixture01 <- setup $ defaultSetup & name .~ "newWallet1"
```

File: test/integration/Test/Integration/Scenario/Transactions.hs [hunk 4]
```diff
@@ -4,12 +4,116 @@ module Test.Integration.Scenario.Transactions
...
+
+        randomBackPhrase2 <- liftIO $ generate arbitrary :: Scenario Context  IO Client.BackupPhrase
+        wallet2 <- successfulRequest $ Client.postWallet $- NewWallet
+                      randomBackPhrase2
+                      noSpendingPassword
+                      NormalAssurance
+                      "NewWallet2"
+                      CreateWallet
+        account2 <- successfulRequest $ Client.getAccounts
+                    $- wallet2 ^. walletId
```
Comment (KtorZ) [inline test/integration/Test/Integration/Scenario/Transactions.hs]: Same as above :+1:

File: test/integration/Test/Integration/Scenario/Transactions.hs [hunk 5]
```diff
@@ -4,12 +4,116 @@ module Test.Integration.Scenario.Transactions
...
+        let recepientWalAddresses = map (\acc -> acc ^. walAddresses)
+                                    $ concat
+                                    $ map (\acc -> acc ^. addresses)
+                                    $ account1 ++ account2
+
+        let customizedDistribution = fromList
+                $ zipWith
+                Client.PaymentDistribution
+                recepientWalAddresses
+                (map (\coin -> Client.WalletCoin $ mkCoin coin) [10,11])
```
Comment (KtorZ) [inline test/integration/Test/Integration/Scenario/Transactions.hs]: We really want to keep the complexity and the internal logic outside of the scenario. The `DSL` already offers:

```hs
defaultDistribution
    :: HasType (NonEmpty Address) s
    => Word64
    -> s
    -> NonEmpty PaymentDistribution
```

which isn't enough in your case, so I'd suggest to add a new:

```hs
customDistribution
    :: NonEmpty (Address, Word64)
    -> NonEmpty PaymentDistribution
```

File: test/integration/Test/Integration/Scenario/Transactions.hs [hunk 6]
```diff
@@ -4,12 +4,116 @@ module Test.Integration.Scenario.Transactions
...
+
+        response <- request $ Client.postTransaction $- Payment
+            (defaultSource fixture)
+            customizedDistribution
+            defaultGroupingPolicy
+            noSpendingPassword
+
+        verify response
+            [ expectTxInHistoryOf (fixture ^. wallet)
+            , expectTxStatusEventually [Creating, Applying, InNewestBlocks, Persisted]
```
Comment (KtorZ) [inline test/integration/Test/Integration/Scenario/Transactions.hs]: I'd rather wait for `InNewestBlocks` here, to make sure the transaction is actually indeed accepted by the nodes.

**Review by KtorZ — CHANGES_REQUESTED** [review]

File: src/Cardano/Wallet/API/V1/Types.hs (hunk 1 above)
Comment (paweljakubas) [reply src/Cardano/Wallet/API/V1/Types.hs]: done

File: src/Cardano/Wallet/Kernel/Transactions.hs (hunk 2 above)
Comment (paweljakubas) [reply src/Cardano/Wallet/Kernel/Transactions.hs]: done

File: test/integration/Test/Integration/Scenario/Transactions.hs (hunk 3 above)
Comment (paweljakubas) [reply test/integration/Test/Integration/Scenario/Transactions.hs]: applied. now it is much easier to read and all imp delegated to DSL

File: test/integration/Test/Integration/Scenario/Transactions.hs (hunk 4 above)
Comment (paweljakubas) [reply test/integration/Test/Integration/Scenario/Transactions.hs]: :+1:

File: test/integration/Test/Integration/Scenario/Transactions.hs (hunk 5 above)
Comment (paweljakubas) [reply test/integration/Test/Integration/Scenario/Transactions.hs]: done

File: test/integration/Test/Integration/Scenario/Transactions.hs (hunk 6 above)
Comment (paweljakubas) [reply test/integration/Test/Integration/Scenario/Transactions.hs]: indeed -> done

File: src/Cardano/Wallet/API/V1/Types.hs [hunk 7]
```diff
@@ -2410,6 +2412,26 @@ instance FromJSON ErrNotEnoughMoney where
...
+msgUtxoNotEnoughFragmented = "Utxo is not enough fragmented to handle the number of outputs of this transaction. Query /api/v1/wallets/{walletId}/statistics/utxos endpoint for more information"
+
+deriveJSON Aeson.defaultOptions ''ErrUtxoNotEnoughFragmented
+
+instance Buildable ErrUtxoNotEnoughFragmented where
+    build (ErrUtxoNotEnoughFragmented missingUtxos _ ) =
+        if missingUtxos == 1 then
+            bprint ("Missing "%build%" utxo to accommodate all outputs of the transaction") missingUtxos
+        else
+            bprint ("Missing "%build%" utxos to accommodate all outputs of the transaction") missingUtxos
```
Comment (KtorZ) [inline src/Cardano/Wallet/API/V1/Types.hs]: Nice effort :smile: 
Pro tip: an `(s)` works marvelously for this kind of things:

```hs
             bprint ("Missing "%build%" utxo(s) to accommodate all outputs of the transaction") missingUtxos 
```

And, in general, as a backend, the accuracy of your grammar on edge-cases is really your last concern ^^"

File: test/unit/Servant/JsendCompliance.hs [hunk 8]
```diff
@@ -20,7 +20,7 @@ import           Servant.Swagger.Internal.Test (props)
 import           Servant.Swagger.Internal.TypeLevel (AddBodyType, AppendList,
                      Every, Nub, TMap)
 import           Test.QuickCheck (Arbitrary (..), Property, counterexample,
-                     (.&&.), (===))
+                     (===))
```
Comment (KtorZ) [inline test/unit/Servant/JsendCompliance.hs:23]: Minor remark: when you fix something like this (which has nothing to do with our PR per se), make it a separate commit. Such that, the history is clearer to understand and, commits can be cherry-picked / reverted easily when they need to. 

Thanks for fixing the warning though :+1:

File: test/integration/Test/Integration/Scenario/Transactions.hs [hunk 9]
```diff
@@ -49,7 +104,7 @@ spec = do
             noSpendingPassword
 
         verify response
-            [ expectWalletError (NotEnoughMoney (ErrAvailableBalanceIsInsufficient 0))
+            [ expectWalletError (UtxoNotEnoughFragmented (Client.ErrUtxoNotEnoughFragmented 1 Client.msgUtxoNotEnoughFragmented))
```
Comment (KtorZ) [inline test/integration/Test/Integration/Scenario/Transactions.hs:107]: :+1: Note that, this is a small "breaking" change and we should notify the frontend folks about this!

File: src/Cardano/Wallet/API/V1/Types.hs [hunk 10]
```diff
@@ -2657,3 +2689,5 @@ instance HasDiagnostic WalletError where
             "msg"
         RequestThrottled{} ->
             "microsecondsUntilRetry"
+        UtxoNotEnoughFragmented{} ->
+            "msg"
```
Comment (KtorZ) [inline src/Cardano/Wallet/API/V1/Types.hs]: Note, by using "msg" as a diagnostic key, we'll end up with an error like:

```json
{    
    "status": "error", 
    "diagnostic": {        
        "msg": {            
            "msg": "...",            
            "missingUtxos": 14        
        }    
    },    
    "message": "UtxoNotEnoughFragmented"
}
```

So that's quite a lot of 'msg / messages' ... May I recommand to change the diagnostic key to `details` and maybe also, to change the `theMsg` by `theHelp` in the `ErrUtxoNotEnoughFragmented` type ? This will generate something more like:

```json
{    
    "status": "error", 
    "diagnostic": {        
        "details": {            
            "help": "...",            
            "missingUtxos": 14        
        }    
    },    
    "message": "UtxoNotEnoughFragmented"
}
```

File: test/integration/Test/Integration/Framework/DSL.hs [hunk 11]
```diff
@@ -265,6 +269,18 @@ defaultDistribution
 defaultDistribution c s = pure $
     PaymentDistribution (WalAddress $ head $ s ^. typed) (WalletCoin $ mkCoin c)
 
+customDistribution
+    :: [Account]
+    -> [Word64]
+    -> NonEmpty PaymentDistribution
+customDistribution accounts coins =
+    let recepientWalAddresses = map (\acc -> acc ^. walAddresses)
+                                $ concatMap (\acc -> acc ^. addresses)
```
Comment (KtorZ) [inline test/integration/Test/Integration/Framework/DSL.hs]: NOTE: `\x -> x ^. myLens === view myLens`, favor functions over operator when passing functions around :+1:

File: test/integration/Test/Integration/Framework/DSL.hs [hunk 12]
```diff
@@ -265,6 +269,18 @@ defaultDistribution
...
+                                $ accounts
+    in fromList $
+       zipWith PaymentDistribution recepientWalAddresses (map (\coin -> WalletCoin $ mkCoin coin) coins)
```
Comment (KtorZ) [inline test/integration/Test/Integration/Framework/DSL.hs]: This may lead to unexpected results if:

- The `accounts` and `coins` list don't have the same number of elements
- The `accounts` have more than one address 

I'd recommend to either check that as invariant and fail hard if the invariant aren't met, or (better IMO), require the caller to provide the a `(Account, Word64)]` himself and only take the first address of each account -- failing if there's none maybe. Or asking directly for `[(Address, Word64)]`

File: test/unit/Servant/JsendCompliance.hs (hunk 8 above)
Comment (KtorZ) [reply test/unit/Servant/JsendCompliance.hs:23]: Also, this happens because `-WError` isn't on in `cardano-wallet.cabal` for the unit tests. So I'd rather turn it on as well while we're at it :)

**Review by KtorZ — CHANGES_REQUESTED** [review]

File: src/Cardano/Wallet/API/V1/Types.hs (hunk 7 above)
Comment (paweljakubas) [reply src/Cardano/Wallet/API/V1/Types.hs]: yeah, a little bit overstretched here :-) corrected

File: test/unit/Servant/JsendCompliance.hs (hunk 8 above)
Comment (paweljakubas) [reply test/unit/Servant/JsendCompliance.hs:23]: reverted the change, and included in another commit (together with switching on of -Werror flag for unit tests)

File: test/integration/Test/Integration/Scenario/Transactions.hs (hunk 9 above)
Comment (paweljakubas) [reply test/integration/Test/Integration/Scenario/Transactions.hs:107]: Will inform Nicola team about the change

File: src/Cardano/Wallet/API/V1/Types.hs (hunk 10 above)
Comment (paweljakubas) [reply src/Cardano/Wallet/API/V1/Types.hs]: done

File: test/integration/Test/Integration/Framework/DSL.hs (hunk 11 above)
Comment (paweljakubas) [reply test/integration/Test/Integration/Framework/DSL.hs]: yes, used `view` instead

File: test/integration/Test/Integration/Framework/DSL.hs (hunk 12 above)
Comment (paweljakubas) [reply test/integration/Test/Integration/Framework/DSL.hs]: done

**Review by KtorZ — APPROVED** [review]

**Comment by KtorZ** [comment]
Nice work @paweljakubas! Thanks. Don't forget to put your ticket in QA after giving a few pointers to @piotr-iohk!
